pkg/controller/attach_detach.go

```diff
@@ -741,45 +723,46 @@ func (r *ReconcileAttachDetach) recoverAzVolumeAttachment(ctx context.Context, r
 		wg.Add(1)
 		go func(azv azdiskv1beta2.AzVolumeAttachment, azvMap *sync.Map) {
 			defer wg.Done()
+			var publishCtx map[string]string
+			if lun, _, err := r.cloudDiskAttacher.GetDiskLun(azv.Spec.VolumeName, azv.Spec.VolumeID, types.NodeName(azv.Spec.NodeName)); err == nil {
+				publishCtx = map[string]string{"LUN": strconv.Itoa(int(lun))}
+			}
 			var targetState azdiskv1beta2.AzVolumeAttachmentAttachmentState
 			updateMode := azureutils.UpdateCRIStatus
-			if azv.Spec.RequestedRole == azdiskv1beta2.ReplicaRole {
-				updateMode = azureutils.UpdateAll
-			}
 			updateFunc := func(obj client.Object) error {
 				var err error
 				azv := obj.(*azdiskv1beta2.AzVolumeAttachment)
-				if azv.Spec.RequestedRole == azdiskv1beta2.ReplicaRole {
-					// conversion logic from v1beta1 to v1beta2 for replicas come here
-					azv.Status.Annotations = azv.ObjectMeta.Annotations
-					azv.ObjectMeta.Annotations = map[string]string{consts.VolumeAttachRequestAnnotation: "CRI recovery"}
-				}
+				azv = updateStatusDetail(azv, publishCtx)
 				// add a recover annotation to the CRI so that reconciliation can be triggered for the CRI even if CRI's current state == target state
 				azv.Status.Annotations = azureutils.AddToMap(azv.Status.Annotations, consts.RecoverAnnotation, recoveryID)
 				if azv.Status.State != targetState {
 					_, err = updateState(azv, targetState, forceUpdate)
 				}
 				return err
 			}
-			switch azv.Status.State {
-			case azdiskv1beta2.Attaching:
-				// reset state to Pending so Attach operation can be redone
-				targetState = azdiskv1beta2.AttachmentPending
-				innerUpdateFunc := updateFunc
-				updateFunc = func(obj client.Object) error {
-					if err := innerUpdateFunc(obj); err != nil {
-						return err
+			if r.isAttached(ctx, azv.Spec.VolumeID, azv.Spec.NodeName) {
+				targetState = azdiskv1beta2.Attached
+			} else {
+				switch azv.Status.State {
+				case azdiskv1beta2.Attaching:
```

Comment:
Sorry, I'm not sure where the `azv.Status.State` has been set when recreating the azvolumeattachment. But is it possible that a primary azvolumeattachment's state is Attached, however, `isAttached` function returns false due to async attach feature? In this case, should we change its state to AttachmentPending as well to redo the attach operation?

Reply:
When we call `recreateAzVolumeAttachment()`, a create event will be triggered and `Reconcile()` can catch the event in this line:
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/448acb3466b2ae888cdf3aa271bab9f911c18aa2/pkg/controller/attach_detach.go#L138
Then `triggerAttach()` will update status.

Reply:
Thanks for bringing up this corner case. To fix it, I will add the check `isAttached()` before `recreateAzVolumeAttachment()`. If `isAttached()` returns false, I will not recreate it.

Reply:
It's a possible scenario for the driver to crash in the middle of an attach operation, in which case it's very possible for the `AzVolumeAttachment.Status.State` and the actual attach state of the disk to be inconsistent (i.e. `AzVolumeAttachment.Status.State` would be `Attaching` when the disk is, in fact, attached.

The driver currently handles this scenario by updating the state from `Attaching` to `AttachmentPending` to restart the attach operation to update the state but given the new feature this PR adds to check the attach state of the disk, we can directly infer and update the state of the `AzVolumeAttachment` CRI at the recovery stage instead of relying on an additional + unnecessary attach operation on an already attached disk.